test/integration/shoots/applications/shoot_app_test.go

```diff
@@ -0,0 +1,332 @@
+		Expect(err).NotTo(HaveOccurred())
+
+		err = shootTestOperations.WaitUntilStatefulSetIsRunning(ctx, "redis-master", helmDeployNamespace, shootTestOperations.K8sShootClient)
+		Expect(err).NotTo(HaveOccurred())
+
+		redisSlaveLabelSelector := labels.SelectorFromSet(labels.Set(map[string]string{
+			"app":  "redis",
+			"role": "slave",
+		}))
+
+		err = shootTestOperations.WaitUntilDeploymentsWithLabelsIsReady(ctx, redisSlaveLabelSelector, helmDeployNamespace, shootTestOperations.K8sShootClient)
+		Expect(err).NotTo(HaveOccurred())
+
+		guestBookParams := struct {
+			HelmDeployNamespace string
+			ShootDNSHost        string
+		}{
+			helmDeployNamespace,
+			fmt.Sprintf("guestbook.ingress.%s", *shootTestOperations.Shoot.Info.Spec.DNS.Domain),
+		}
+
+		// Parse the guestbook app template
+		var writer bytes.Buffer
+		err = guestBooktpl.Execute(&writer, guestBookParams)
+		Expect(err).NotTo(HaveOccurred())
+
+		// Apply the guestbook app resources to shoot
+
+		manifestReader := kubernetes.NewManifestReader(writer.Bytes())
+		err = shootTestOperations.K8sShootClient.Applier().ApplyManifest(context.TODO(), manifestReader)
+		Expect(err).NotTo(HaveOccurred())
+
+		// define guestbook app urls
+		guestBookAppURL := fmt.Sprintf("http://guestbook.ingress.%s", *shootTestOperations.Shoot.Info.Spec.DNS.Domain)
+		pushString := fmt.Sprintf("foobar-%s", shootTestOperations.Shoot.Info.Name)
+		pushUrl := fmt.Sprintf("%s/rpush/guestbook/%s", guestBookAppURL, pushString)
+		pullUrl := fmt.Sprintf("%s/lrange/guestbook", guestBookAppURL)
+
+		// Check availability fo the guestbookapp
+		err = shootTestOperations.WaitUntilGuestbookAppIsAvailable(ctx, []string{guestBookAppURL, pushUrl, pullUrl})
+		Expect(err).NotTo(HaveOccurred())
+
+		// Push foobar-<shoot-name> to the guestbook app
+		_, err = shootTestOperations.HTTPGet(ctx, pushUrl)
+		Expect(err).NotTo(HaveOccurred())
+
+		// Pull foobar
+		pullResponse, err := shootTestOperations.HTTPGet(ctx, pullUrl)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(pullResponse.StatusCode).To(Equal(http.StatusOK))
+
+		responseBytes, err := ioutil.ReadAll(pullResponse.Body)
+		Expect(err).NotTo(HaveOccurred())
+
+		// test if foobar-<shoot-name> was pulled successfully
+		bodyString := string(responseBytes)
+		Expect(bodyString).To(ContainSubstring(fmt.Sprintf("foobar-%s", shootTestOperations.Shoot.Info.Name)))
+
+		By("Guestbook app was deployed successfully!")
+
+	}, GuestbookAppTimeout)
+
+	CIt("Dashboard should be available", func(ctx context.Context) {
+		err := shootTestOperations.DashboardAvailable(ctx)
+		Expect(err).NotTo(HaveOccurred())
+	}, DashboardAvailableTimeout)
+
+	Context("Network Policy Testing", func() {
+		var (
+			NetworkPolicyTimeout = 1 * time.Minute
+
+			ExecNCOnAPIServer = func(ctx context.Context, host, port string) error {
+				return shootTestOperations.PodExecByLabel(ctx, apiserverLabels, APISERVER,
```

Comment:
How can we actually be sure that we always have the same pod to `exec` on? Your code always gets the first running pod that matches the labels, but if we, at some point in the future, decided to always scale the api server to e.g. at least 2 you cannot guarantee it since the order of pods in the retrieved list is nondeterministic (at least API wise).

Reply:
True, but the goal of this test is to test network policies which work based on labels. What this test confirms is that a pod with labels x & y can reach to set{pod1,pod2,pod3}. What you might want to do in the future is a connectivity test from all pods in a deployment to all other pods. In that case we can build a util `ExecOnAllWithLabel` that does this.